Integrating tests for E2

smartie2076 opened the pull request:
- [x] Create file for tests (tests_*.py)
- [x] Create docstrings describing functionalities in module
- [ ] Add names of test functions
- [ ] Define pytests
- [ ] Change CHANGELOG.md

smartie2076 commented:
Hi @ursulaelmir! Do you think you could take over this PR from me? It again deals with economic functions and stuff.

ursulaelmir commented:
> Hi @ursulaelmir! Do you think you could take over this PR from me? It again deals with economic functions and stuff.

yes will check it out :)

ursulaelmir commented:
@smartie2076 

A couple of questions regarding the get_costs function in E2.

1. In line 89, why are costs_energy only being added to cost_om? Isn't cost_om part of costs_total?

2. In line 98, we are looking if lifetime_opex_var is in the list, however we are using lifetime_opex_fix instead. Should we look for that instead in the if condition?

3. In line 97, why are we still looking for annual_total_flow if we won't use it for the rest of the calculations?

smartie2076 commented:
Hi Ursula! 

Thanks for writing down the issue, that way I can still answer them. Can you next time also link the lines that you are referring to? For that, you go into a file on github, and then to the appropirate line, click on the number and copy the link from the link bar above.

I added links below. Can you check if I added the correct lines?

Q1. [In line 89](https://github.com/rl-institut/mvs_eland/blob/feature/tests_E2/src/E2_economics.py#L89), why are costs_energy only being added to cost_om? Isn't cost_om part of costs_total?

Q2. [In line 98](https://github.com/rl-institut/mvs_eland/blob/feature/tests_E2/src/E2_economics.py#L98), we are looking if lifetime_opex_var is in the list, however we are using lifetime_opex_fix instead. Should we look for that instead in the if condition?

Q3. [In line 97](https://github.com/rl-institut/mvs_eland/blob/feature/tests_E2/src/E2_economics.py#L97), why are we still looking for annual_total_flow if we won't use it for the rest of the calculations?

So... I was about to answer each of your questions. But I would do so much different by now, and describing it all would actually mean doing this PR myself. 
So then I realized, that you simply have to change perspective to tackle this PR. Assume the code is wrong. What you basically have to do now is not "I have code, and I believe in it and test it", but "I write tests and make sure the code I am coding does what I want it to" (test-driven development, see #278 ). 

You can work with the code that is preexisting, rearrange the functions / create new ones so that it is easier to read, or start anew. But you should proceed as following:
1. Do a cost calculation by hand
2. Write tests for those calculations which fail (eg. because they link to empty functions)
3. Develop the code so that the calculations do not fail anymore

smartie2076 commented:
You should test dictionaries somewhat like [this](https://github.com/rl-institut/mvs_eland/pull/257/files#diff-20a5d5c7986c1038f0a85f406787f1caR174).

smartie2076 commented:
Hi @ursulaelmir ! Can you update the changes that you did locally?

smartie2076 commented:
Hi @ursulaelmir ! I wanted to add on to the code I pushed yesterday with one of the tests that I wrote (for reference):

```
def test_if_simulation_results_added_to_dict_values():
    D0.run_oemof(dict_values)
    for k in ("label", "objective_value", "simulation_time"):
        assert (k in dict_values["simulation_results"].keys())
```

Bachibouzouk commented on tests/test_E2_economics.py:
```diff
@@ -18,8 +18,11 @@
+    """
 
+    Tests whether the function get_costs is adding all the calculated costs to dict_asset
+    """
```
Just a suggestion for docstring format convention by [PEP257](https://www.python.org/dev/peps/pep-0257/) ^^

smartie2076 commented on tests/test_E2_economics.py:
```diff
@@ -0,0 +1,54 @@
+    assert "new_cost" in dict_asset
+
+
+def test_all_list_in_dict():
```
> def test_all_list_in_dict():

I am not quite sure what this test does. Also, if you have two parts of one tests where you once check "True" and once "False", maybe put it into the function name?

ursulaelmir replied:
I defined two lists and was trying to check if calling the function and looking for the list items in the dict works fine.

smartie2076 commented on tests/test_E2_economics.py:
```diff
@@ -0,0 +1,54 @@
+    assert "new_cost" in dict_asset
+
+
+def test_all_list_in_dict():
+    """Tests whether looking for list items in dict_asset is plausible."""
+    list_true = ["annual_total_flow", "optimizedAddCap"]
+    boolean = E2.all_list_in_dict(dict_asset, list_true)
+    assert boolean == True
+
+    list_false = ["flow", "optimizedAddCap"]
+    boolean = E2.all_list_in_dict(dict_asset, list_false)
+    assert boolean == False
```
How about this then?
```suggestion
def test_all_list_in_dict_passes_as_all_keys_included():
    """Tests whether looking for list items in dict_asset is plausible."""
    list_true = ["annual_total_flow", "optimizedAddCap"]
    boolean = E2.all_list_in_dict(dict_asset, list_true)
    assert boolean == True
    
def test_all_list_in_dict_fails_due_to_not_included_keys():
    """Tests whether looking for list items in dict_asset is plausible."""
    list_false = ["flow", "optimizedAddCap"]
    boolean = E2.all_list_in_dict(dict_asset, list_false)
    assert boolean == False
```
